network/src/testutils/test_node.rs

```diff
@@ -0,0 +1,313 @@
+        handle: InboundNetworkHandle,
+    );
+
+    /// For sending to other nodes
+    fn other_inbound_handle(&self, peer_network_id: PeerNetworkId) -> InboundNetworkHandle;
+
+    /// For receiving messages from other nodes
+    fn outbound_handle(&mut self, network_id: NetworkId) -> &mut OutboundMessageReceiver;
+
+    fn peer_metadata_storage(&self) -> &PeerMetadataStorage;
+
+    fn peer_network_ids(&self) -> &HashMap<NetworkId, PeerNetworkId>;
+
+    fn peer_network_id(&self, network_id: NetworkId) -> PeerNetworkId {
+        *self.peer_network_ids().get(&network_id).unwrap_or_else(|| {
+            panic!(
+                "Expected network {} to exist on node {}",
+                network_id,
+                self.node_id()
+            )
+        })
+    }
+
+    fn network_ids(&self) -> Vec<NetworkId> {
+        self.peer_network_ids().keys().copied().collect()
+    }
+
+    fn role(&self) -> RoleType {
+        match self.node_id().node_type {
+            NodeType::Validator => RoleType::Validator,
+            _ => RoleType::FullNode,
+        }
+    }
+
+    fn peer_role(&self) -> PeerRole {
+        match self.node_id().node_type {
+            NodeType::Validator => PeerRole::Validator,
+            NodeType::ValidatorFullNode => PeerRole::ValidatorFullNode,
+            NodeType::PublicFullNode => PeerRole::Unknown,
+        }
+    }
+
+    /// Connects a node to another node.  The other's inbound handle must already be added.
+    fn connect(&self, network_id: NetworkId, metadata: ConnectionMetadata) {
+        let self_peer_id = self.peer_network_id(network_id).peer_id();
+        let self_metadata = self.conn_metadata(network_id, ConnectionOrigin::Inbound, None);
+        let remote_peer_id = metadata.remote_peer_id;
+        // Tell the other node it's good to send to the connected peer now
+        self.other_inbound_handle(PeerNetworkId::new(network_id, remote_peer_id))
+            .connect(remote_peer_id, network_id, self_metadata);
+        // Then connect us
+        self.node_inbound_handle(network_id)
+            .connect(self_peer_id, network_id, metadata);
+    }
+
+    /// Disconnects a node from another node
+    fn disconnect(&self, network_id: NetworkId, metadata: ConnectionMetadata) {
+        let self_peer_id = self.peer_network_id(network_id).peer_id();
+        let self_metadata = self.conn_metadata(network_id, ConnectionOrigin::Inbound, None);
+        let remote_peer_id = metadata.remote_peer_id;
+        // Tell the other node it's disconnected
+        self.other_inbound_handle(PeerNetworkId::new(network_id, remote_peer_id))
+            .disconnect(remote_peer_id, network_id, self_metadata);
+        // Then disconnect us
+        self.node_inbound_handle(network_id)
+            .disconnect(self_peer_id, network_id, metadata);
+    }
+
+    /// Find a common network between nodes, they're sorted in priority order
+    fn find_common_network(&self, other: &Self) -> Option<NetworkId> {
```

Comment:
Unless we plan on overriding this, I don't think it should be in the trait. A helper function seems more natural?

Reply:
Would you accept an extension trait?

Basically, I don't want to have to pass in the node on all of these functions.  The only helper functions I created to pass in were async, so I didn't have to make async traits